### PR TITLE
[AMDGPU] Defaults for missing dimensions in SYCL required wg size

### PR DESCRIPTION
**Pull request by jchlanda**
SYCL allows for required work group to be partially specified (i.e. not all 3 dimensions):
* https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes

This fails AMDGPU's attribute verification. The patch aims to provide the default values for missing dimensions when dealing with SYCL kernels. Rather than modifying the module's metadata it uses internal data to padd missing values.

### CODE REVIEW

**Comment by jchlanda** [comment]
Friendly ping @arsenm

File: llvm/lib/Target/AMDGPU/AMDGPUHSAMetadataStreamer.cpp [hunk 1]
```diff
@@ -173,14 +173,18 @@ std::string MetadataStreamerMsgPackV4::getTypeName(Type *Ty,
 }
 
 msgpack::ArrayDocNode
-MetadataStreamerMsgPackV4::getWorkGroupDimensions(MDNode *Node) const {
+MetadataStreamerMsgPackV4::getWorkGroupDimensions(const Function &Func,
+                                                  MDNode *Node) const {
   auto Dims = HSAMetadataDoc->getArrayNode();
-  if (Node->getNumOperands() != 3)
+  if (Node->getNumOperands() != 3 && !Func.hasFnAttribute("sycl-module-id"))
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUHSAMetadataStreamer.cpp:179]: The attribute check should not be necessary. The metadata should have a standalone interpretation and not depend on any other attributes

File: llvm/test/CodeGen/AMDGPU/required_work_group_size_sycl.ll [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+; Make sure that SYCL kernels with less than 3 dimensions specified in required
+; work group size, have those dimensions padded up with 1.
+
+; CHECK-LABEL: .name:           sycl_kernel_1dim
+; CHECK:    .reqd_workgroup_size:
+; CHECK-NEXT:      - 3
+; CHECK-NEXT:      - 1
+; CHECK-NEXT:      - 1
+define weak_odr protected amdgpu_kernel void @sycl_kernel_1dim() #1 !reqd_work_group_size !0 {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/required_work_group_size_sycl.ll:11]: Don't need the weak_odrs

File: llvm/test/CodeGen/AMDGPU/required_work_group_size_sycl.ll [hunk 3]
```diff
@@ -0,0 +1,42 @@
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx90a -verify-machineinstrs < %s | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/required_work_group_size_sycl.ll:1]: Don't bother with -verify-machineinstrs, it's unlikely to catch anything here and will be caught in EXPENSIVE_CHECKS builds

**Review by arsenm — COMMENTED** [review]
I still think we need to have one single spec for the metadata and not modify based on any is-this-language type annotations. The different producers and consumers should just agree on the operand constraints of the metadata

**Comment by jchlanda** [comment]
Closing this PR, we've decide to handle it internally in SYCL (https://github.com/intel/llvm/pull/13600). It aligns with OpenCL's and always pads missing dimensions with `1`.

**Pull request closed without merging**